### DISCUSSION TITLE
Remove metric series where Label Values are no longer in use. If Label Values are Dynamic, is there a way to remove the previous labels from being reported?

### DISCUSSION THREAD

**Discussion by arjunbakshi245** [prometheus/client_golang#1299] [Q&A] [answered]
Hi team,
I am exposing metrics for a custom struct in golang. Each object of the struct in named in a particular manner uniquely and dynamically so as to report metrics correctly for each name/object. This name is passed as the label value of the metric(For e.g. `struct_free_size(name="moduleName: UUID4u133678")`). Naming would have been similar to `"moduleName:" + obj.UUID`. However, some objects of this struct after some operation no longer exist, hence it makes no sense to continuously report metrics for them. 
I want to know if there is a way to remove all the labels of a metric through a method here or something similar. 

I have tried to maintain the labels map I want to delete. But the .Delete(labels) deletes the metrics before exporting and hence I don't see anything on the endpoint. I want to delete the labels after I have sent them, so that in the next poll they do not show up.
I have a registered metric and I am iterating over a list to emit metrics for each entity in the list. at time `t1` the list is `[{"a",2},{"b",3},{"c",10}]` then let's say the output is
```
# Name .....
# Help ......
metric_name{name = "a"} 2
metric_name{name = "b"} 3
metric_name{name = "c"} 10
```
Now after some time, let's say at time `t2` the list we are iterating over is `[{"a",4},{"b",9}]`. But now the `{"c",10}` still shows on the endpoint? I want the output to be as the one Below.
```
# Name .....
# Help ......
metric_name{name = "a"} 4
metric_name{name = "b"} 9
```

**Comment by bwplotka** [comment, marked as answer]
Hi!

Yes, you can use [`Reset`](https://pkg.go.dev/github.com/prometheus/client_golang@v1.16.0/prometheus#MetricVec.Reset) method on all Vec* metric types (e.g. GaugeVec). There are also [`Delete*`](https://pkg.go.dev/github.com/prometheus/client_golang@v1.16.0/prometheus#MetricVec.Delete) methods you can use. (: 

BTW: However, looking on your metric, you might want to look at profiling (if interested in aggregation of performance metrics like mem/CPU per struct/ops). That's ofc depending what do you want to achieve (:

**Reply by arjunbakshi245** [reply]
Hi @bwplotka . I want to know a little more about the Delete methods. More particularly about the **DeleteLabelValues**. I have a metric with only one label. So I have used it as follows `entity.metrics.entityFreeSize.DeleteLabelValues(entity.name)`(here the name of the entity is the value of the label). I want to know what exactly this does. I think it removes the metric from the endpoint on which the metrics are emitted, however, I am not sure if the timeseries created by it initially still exists in memory. It may be deleted from the end point but is it possible that somewhere down, the metric is still listening for this label(supposed to be deleted), but not emitting anything because of the **DeleteLabelValues** method. In short, I want to confirm if there is no **memory leak**.
